### PR TITLE
Kernel: Split the mount syscall into a bunch of more-atomic syscalls (part 1)

### PR DESCRIPTION
**Pull request by supercomputer7**
Related to #16299.

The idea is to have a better control on the `mount` procedure in general, so one can think of a userspace program that issues multiple syscalls which each syscall does an "atomic" operation on a `MountFile` object (which is our "context" file descriptor here, in contrast to linux which has a real different `context` struct), so ideally userspace can understand on what things might fail during the procedure by checking for errors after each syscall that happened.

This is also vital for allowing flexible expansion in the mount procedure to support fs-specific flags, which is what #16299 is about.
This also fixes some issues we had with the very-complicated `mount` syscall - for example, the `mount` syscall acquired the big lock, and the new syscalls don't do that. Other fundamental problems are resolved, as being described in the commit messages.

cc @ADKaster

### CODE REVIEW

**Comment by supercomputer7** [comment]
I figured that there's a flaw in this design - apparently two processes could use a single `BlockDevice` in a race condition to create two separate `FileSystem` objects to be mounted, which is quite bad.
I should look into how to fix this, but until I solve this, we should not merge this.

**Comment by supercomputer7** [comment]
Until I figure out how to fix this design, let's close this.

**Comment by supercomputer7** [comment]
Let's try this again.

File: Kernel/API/FileSystem/MountSpecificFlags.h
Comment (kleinesfilmroellchen) [inline Kernel/API/FileSystem/MountSpecificFlags.h:11]: Any specific reason why this is packed? Looking at the contents it looks like it won't have any padding anyways, and unaligned pointers are only asking for trouble on ARM-64...

File: Kernel/FileSystem/Initializer.h
Comment (kleinesfilmroellchen) [inline Kernel/FileSystem/Initializer.h:26]: It appears the boolean value argument is missing here, or is this function always going to flip the key's value? If so, I don't think that's a good design, since there may be varying defaults, some false, some true.

File: Kernel/FileSystem/MountFile.cpp
Comment (kleinesfilmroellchen) [inline Kernel/FileSystem/MountFile.cpp:25]: Don't use NOTE, since comments are notes anyways :^)

File: Kernel/FileSystem/MountFile.cpp
Comment (kleinesfilmroellchen) [inline Kernel/FileSystem/MountFile.cpp:64]: Any specific reason for these limitations other than "do not OOM the kernel"? If so, we already copied the data from user memory to kernel memory, so I'm not sure whether this helps much.

File: Kernel/FileSystem/MountFile.cpp
Comment (kleinesfilmroellchen) [inline Kernel/FileSystem/MountFile.cpp:70]: no NOTE

File: Kernel/FileSystem/MountFile.cpp
Comment (kleinesfilmroellchen) [inline Kernel/FileSystem/MountFile.cpp:99]: Can be removed since there's a `return EINVAL` for the default case already.

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]
50% Partial review

File: Kernel/FileSystem/MountFile.cpp
Comment (supercomputer7) [reply Kernel/FileSystem/MountFile.cpp:25]: I already used NOTE a lot in the past (and other people did that too) and I see no problems with this pattern. This helps differentiate this from FIXME notes, for example.

File: Kernel/FileSystem/MountFile.cpp
Comment (supercomputer7) [reply Kernel/FileSystem/MountFile.cpp:70]: I already used NOTE a lot in the past (and other people did that too) and I see no problems with this pattern. This helps differentiate this from FIXME notes, for example.

File: Kernel/FileSystem/MountFile.cpp
Comment (supercomputer7) [reply Kernel/FileSystem/MountFile.cpp:64]: This is not correct. The actual copying of the string happens in line 92 of that file, for an `ASCIIString` type. We check that the `ASCIIString` is not very long, but the `mount_specific_data` struct is always going to be 28 bytes, exactly.

Comment (kleinesfilmroellchen) [reply Kernel/FileSystem/MountFile.cpp:64]: Still, I'm not sure how these limits help anyone. We've had to struggle with plenty of dumb file formats and APIs just because someone thought that `<insert number here>` characters ought to be enough.

File: Kernel/FileSystem/VirtualFileSystem.cpp
Comment (kleinesfilmroellchen) [inline Kernel/FileSystem/VirtualFileSystem.cpp:297]: Should that be a FIXME? Seems like we're missing a smart pointer here...

File: Kernel/API/FileSystem/MountSpecificFlags.h
Comment (GMTA) [inline Kernel/API/FileSystem/MountSpecificFlags.h:11]: We use camel case for struct names

File: Kernel/API/FileSystem/MountSpecificFlags.h
Comment (GMTA) [inline Kernel/API/FileSystem/MountSpecificFlags.h:2]: Just a nit, but you can update all these to `2023` ;-)

File: Kernel/API/FileSystem/MountSpecificFlags.h
Comment (GMTA) [inline Kernel/API/FileSystem/MountSpecificFlags.h:14]: Can't you use `ValueType` directly here after marking it as an  `u32`?

File: Kernel/FileSystem/DevPtsFS/FileSystem.cpp
Comment (GMTA) [inline Kernel/FileSystem/DevPtsFS/FileSystem.cpp:16]: I think we have `ReadonlyBytes` as the default alias for `Span<u8 const>`

File: Kernel/FileSystem/MountFile.cpp
Comment (GMTA) [inline Kernel/FileSystem/MountFile.cpp:30]: What's the magic value `4096`? Is it `PAGESIZE`?

File: Kernel/FileSystem/MountFile.cpp
Comment (GMTA) [reply Kernel/FileSystem/MountFile.cpp:64]: I agree with @kleinesfilmroellchen - if we have some limits, and they are necessary, then we should either explain them or make them easily configurable by using constants somewhere.

File: Kernel/FileSystem/MountFile.cpp
Comment (GMTA) [inline Kernel/FileSystem/MountFile.cpp:73]: Please use C++-style casts everywhere (and yeah, I know that `(FlatPtr)` is used a lot in current code in `Kernel/`)

File: Kernel/FileSystem/MountFile.cpp
Comment (GMTA) [inline Kernel/FileSystem/MountFile.cpp:77]: `ValueType::None` rubs me the wrong way if the value is communicated as a boolean. Maybe call it `ValueType::Boolean` instead?

File: Kernel/FileSystem/VirtualFileSystem.h
Comment (GMTA) [inline Kernel/FileSystem/VirtualFileSystem.h:128]: Why do we go from a spinlock to a mutex? I don't understand the relation to the rest of this commit.

File: Kernel/Storage/StorageManagement.cpp
Comment (GMTA) [inline Kernel/Storage/StorageManagement.cpp:459]: Same here (magic `4096` value)

File: Kernel/Storage/StorageManagement.cpp
Comment (GMTA) [inline Kernel/Storage/StorageManagement.cpp:460]:
```cpp
mount_specific_data.fill(0);
```

File: Userland/Libraries/LibC/unistd.cpp
Comment (GMTA) [inline Userland/Libraries/LibC/unistd.cpp:998]: You can drop the braces for single line if blocks (same for previous `if`)

File: Kernel/FileSystem/MountFile.cpp
Comment (supercomputer7) [reply Kernel/FileSystem/MountFile.cpp:25]: After the discussion in the discord server I think it's fine to let it be this way.

File: Kernel/FileSystem/MountFile.cpp
Comment (supercomputer7) [reply Kernel/FileSystem/MountFile.cpp:70]: After the discussion in the discord server I think it's fine to let it be this way.

File: Kernel/FileSystem/VirtualFileSystem.cpp
Comment (supercomputer7) [reply Kernel/FileSystem/VirtualFileSystem.cpp:297]: We can't use a smart pointer (i.e. `OwnPtr`) here. There was a discussion a few months ago about using Intrusive lists with `OwnPtr` and the conclusion was that a big change to how either the IntrusiveList class works, or introducing a new one (called `IntrusiveListOwnObjects` or something like this) to do that.
Also, this code seems to be new, but it's just that the diff is a bit screwed up and this code was already there.

File: Kernel/API/FileSystem/MountSpecificFlags.h
Comment (supercomputer7) [reply Kernel/API/FileSystem/MountSpecificFlags.h:2]: I think I wrote this in 2022, so I will just add 2023 too.

File: Kernel/FileSystem/MountFile.cpp
Comment (supercomputer7) [reply Kernel/FileSystem/MountFile.cpp:30]: Yes it is, I should use that.

File: Kernel/FileSystem/MountFile.cpp
Comment (supercomputer7) [reply Kernel/FileSystem/MountFile.cpp]: Or we could have both - explaining with a comment and having configurable constants :)

File: Kernel/FileSystem/VirtualFileSystem.h
Comment (supercomputer7) [reply Kernel/FileSystem/VirtualFileSystem.h:128]: I will explain this with a comment and in the commit message.

**Comment by supercomputer7** [comment]
Once this is in our tree, we can re-open #16299.

**Review by kleinesfilmroellchen — APPROVED** [review]

**Comment by supercomputer7** [comment]
Rebased so this doesn't have conflicts now.

**Review by kleinesfilmroellchen — APPROVED** [review]

**Review by gmta — APPROVED** [review]

**Pull request merged**
